allennlp/data/semparse/worlds/nlvr_world.py

```diff
@@ -0,0 +1,436 @@
+        shape_counts: Dict[str, int] = defaultdict(int)
+        for _shape in all_shapes:
+            shape_counts[_shape] += 1
+        if max(shape_counts.values()) <= 1:
+            return set()
+        most_frequent_shape = max(all_shapes, key=shape_counts.get)
+        return cls._filter(objects_set, cls._get_single_shape, most_frequent_shape, operator.eq)
+
+    @classmethod
+    def touch_bottom(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, lambda x: x.get_attribute("y_loc"), 0, operator.eq)
+
+    @classmethod
+    def touch_left(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, lambda x: x.get_attribute("x_loc"), 0, operator.eq)
+
+    @classmethod
+    def touch_top(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, lambda x: x.get_attribute("y_loc") + x.get_attribute("size"),
+                           100, operator.eq)
+
+    @classmethod
+    def touch_right(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, lambda x: x.get_attribute("x_loc") + x.get_attribute("size"),
+                           100, operator.eq)
+
+    @classmethod
+    def touch_wall(cls, objects_set: Set[Object]) -> Set[Object]:
+        return_set: Set[Object] = set()
+        return return_set.union(cls.touch_top(objects_set), cls.touch_left(objects_set),
+                                cls.touch_right(objects_set), cls.touch_bottom(objects_set))
+
+    @classmethod
+    def touch_corner(cls, objects_set: Set[Object]) -> Set[Object]:
+        return_set: Set[Object] = set()
+        return return_set.union(cls.touch_top(objects_set).intersection(cls.touch_right(objects_set)),
+                                cls.touch_top(objects_set).intersection(cls.touch_left(objects_set)),
+                                cls.touch_bottom(objects_set).intersection(cls.touch_right(objects_set)),
+                                cls.touch_bottom(objects_set).intersection(cls.touch_left(objects_set)))
+
+    @classmethod
+    def small(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, lambda x: x.get_attribute("size"), 10, operator.eq)
+
+    @classmethod
+    def medium(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, lambda x: x.get_attribute("size"), 20, operator.eq)
+
+    @classmethod
+    def big(cls, objects_set: Set[Object]) -> Set[Object]:
+        return cls._filter(objects_set, lambda x: x.get_attribute("size"), 30, operator.eq)
+
+    @staticmethod
+    def negate_filter(filter_function: Callable[[Set[Object]], Set[Object]],
+                      objects_set: Set[Object]) -> Set[Object]:
+        # Negate an object filter.
+        return objects_set.difference(filter_function(objects_set))
+
+    @staticmethod
+    def assert_equals(actual_attribute: AttributeType, target_attribute: AttributeType) -> bool:
```

Comment:
Will this ever get a singleton set as its first argument?

Reply:
It's never a set. It's always either a string or an int.